File: src/com/quran/labs/androidquran/database/BookmarksDBHelper.java
package com.quran.labs.androidquran.database;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.data.QuranInfo;

class BookmarksDBHelper extends SQLiteOpenHelper {

   private static final String TAG = "BookmarksDBHelper";
   
	private static final String DB_NAME = "bookmarks.db";
	private static final int DB_VERSION = 2;
	
   private Context cx;
   /*
	static class AyahTable {
		static final String TABLE_NAME = "ayah_bookmarks";
		static final String ID = "_id";
		static final String PAGE = "page";
		static final String SURA = "sura";
		static final String AYAH = "ayah";
		static final String BOOKMARKED = "bookmarked";
		static final String NOTES = "notes";
	}
	
	private static final String CREATE_AYAH_TABLE =
			" create table " + AyahTable.TABLE_NAME + " (" + 
					AyahTable.ID + " INTEGER PRIMARY KEY, " +
					AyahTable.PAGE + " INTEGER NOT NULL, " +
					AyahTable.SURA + " INTEGER NOT NULL, " +
					AyahTable.AYAH + " INTEGER NOT NULL, " +
					AyahTable.BOOKMARKED + " INTEGER NOT NULL, " +
					AyahTable.NOTES + " TEXT);";
	
	static class PageTable {
		static final String TABLE_NAME = "page_bookmarks";
		static final String ID = "_id";
		static final String BOOKMARKED = "bookmarked";
	}
	
	private static final String CREATE_PAGE_TABLE =
			" create table " + PageTable.TABLE_NAME + " (" + 
					PageTable.ID + " INTEGER PRIMARY KEY, " +
					PageTable.BOOKMARKED + " INTEGER NOT NULL);";
	*/
	static class BookmarksTable {
		static final String TABLE_NAME = "bookmarks";
		static final String ID = "_id";
		static final String NAME = "name";
		static final String TYPE = "type";
		static final String DESCRIPTION = "description";
      static final int TYPE_BOOKMARK = 0;
      static final int TYPE_COLLECTION = 1;
	}
	
	private static final String CREATE_BOOKMARKS_TABLE=
			" create table " + BookmarksTable.TABLE_NAME + " (" + 
					BookmarksTable.ID + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
					BookmarksTable.TYPE + " INTEGER NOT NULL, " +
					BookmarksTable.NAME + " TEXT NOT NULL, " +
					BookmarksTable.DESCRIPTION + " TEXT);";
	
	static class BookmarkMapTable {
		static final String TABLE_NAME = "bookmark_map";
		static final String ID = "_id";
		static final String BOOKMARK_ID = "bookmark_id";
		static final String AYAH_ID = "ayah_id";
		static final String TYPE = "type";
		static final int TYPE_AYAH = 0;
		static final int TYPE_PAGE = 1;
	}
	
	private static final String CREATE_BOOKMARK_MAP_TABLE=
			" create table " + BookmarkMapTable.TABLE_NAME + " (" + 
					BookmarkMapTable.ID + " INTEGER PRIMARY KEY, " +
					BookmarkMapTable.BOOKMARK_ID + " INTEGER NOT NULL, " +
					BookmarkMapTable.AYAH_ID + " INTEGER NOT NULL, " +
					BookmarkMapTable.TYPE + " INTEGER NOT NULL DEFAULT 0);";
	
	public BookmarksDBHelper(Context context) {
		super(context, DB_NAME, null, DB_VERSION);
		this.cx = context;
	}
	
	@Override
	public void onCreate(SQLiteDatabase db) {
//		db.execSQL(CREATE_AYAH_TABLE);
//		db.execSQL(CREATE_PAGE_TABLE);
		db.execSQL(CREATE_BOOKMARKS_TABLE);
		db.execSQL(CREATE_BOOKMARK_MAP_TABLE);
      createSampleBookmarks(db);
	}

	@Override
	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
      Log.i(TAG, "Upgrading database from version "+oldVersion+" to version "+newVersion);
	   if (oldVersion < 2) {
	      upgradeToVer2(db);
	   }
	}

   private void upgradeToVer2(SQLiteDatabase db) {
      db.execSQL("DROP TABLE IF EXISTS tags");
      db.execSQL("DROP TABLE IF EXISTS ayah_tag_map");
      onCreate(db);
      copyOldBookmarks(db);
   }
   
   private void createSampleBookmarks(SQLiteDatabase db) {
      try {
         db.execSQL("INSERT INTO bookmarks VALUES (1, 0, ?, null)",
               new String[] {cx.getString(R.string.sample_bookmark_reading)});
         db.execSQL("INSERT INTO bookmarks VALUES (2, 1, ?, null)",
               new String[] {cx.getString(R.string.sample_bookmark_duas)});
         final String s = "INSERT INTO bookmark_map (bookmark_id, ayah_id, type)";
         db.execSQL(s+" VALUES (1, 1, 1)"); // Daily Reading (pg 1)
         db.execSQL(s+" VALUES (2, 301, 0)"); // Quranic Duas (ayah 3:16)
         db.execSQL(s+" VALUES (2, 2920, 0)"); // Quranic Duas (ayah 25:65)
      } catch (Exception e) {
         Log.e(TAG, "Failed to create sample bookmarks", e);
      }
   }
   
   private void copyOldBookmarks(SQLiteDatabase db) {
      try {
         db.execSQL("INSERT INTO bookmarks VALUES (3, 1, ?, null)",
               new String[] {cx.getString(R.string.sample_bookmark_uncategorized)});
         // Copy over ayah bookmarks
         db.execSQL("INSERT INTO bookmark_map (bookmark_id, ayah_id, type) " +
         		"SELECT 3, _id, 0 FROM ayah_bookmarks WHERE bookmarked=1");
         // Copy over page bookmarks
         Cursor c = db.rawQuery("SELECT _id FROM page_bookmarks WHERE bookmarked=1", null);
         while (c.moveToNext()) {
            int ayahId = QuranInfo.getAyahId(c.getInt(0));
            db.execSQL("INSERT INTO bookmark_map (bookmark_id, ayah_id, type) " +
            		"VALUES (3, "+ayahId+", 1)");
         }
         c.close();
      } catch (Exception e) {
         Log.e(TAG, "Failed to copy old bookmarks", e);
      }
   }

}

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBHelper.kt
package com.quran.labs.androidquran.database

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.quran.labs.androidquran.data.Constants
import com.quran.labs.androidquran.util.QuranSettings
import timber.log.Timber
import java.lang.Exception

class BookmarksDBHelper private constructor(context: Context) :
  SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

  internal object BookmarksTable {
    const val TABLE_NAME = "bookmarks"
    const val ID = "_ID"
    const val SURA = "sura"
    const val AYAH = "ayah"
    const val PAGE = "page"
    const val ADDED_DATE = "added_date"
  }

  internal object TagsTable {
    const val TABLE_NAME = "tags"
    const val ID = "_ID"
    const val NAME = "name"
    const val ADDED_DATE = "added_date"
  }

  internal object BookmarkTagTable {
    const val TABLE_NAME = "bookmark_tag"
    const val ID = "_ID"
    const val BOOKMARK_ID = "bookmark_id"
    const val TAG_ID = "tag_id"
    const val ADDED_DATE = "added_date"
  }

  internal object LastPagesTable {
    const val TABLE_NAME = "last_pages"
    const val ID = "_ID"
    const val PAGE = "page"
    const val ADDED_DATE = "added_date"
  }

  companion object {
    private const val DB_NAME = "bookmarks.db"
    private const val DB_VERSION = 3

    const val QUERY_BOOKMARKS =
      "SELECT " + BookmarksTable.TABLE_NAME + "." + BookmarksTable.ID + ", " +
          BookmarksTable.TABLE_NAME + "." + BookmarksTable.SURA + ", " +
          BookmarksTable.TABLE_NAME + "." + BookmarksTable.AYAH + "," +
          BookmarksTable.TABLE_NAME + "." + BookmarksTable.PAGE + ", " +
          "strftime('%s', " + BookmarksTable.TABLE_NAME + "." + BookmarksTable.ADDED_DATE + ")" +
          ", " +
          BookmarkTagTable.TABLE_NAME + "." + BookmarkTagTable.TAG_ID +
          " FROM " +
          BookmarksTable.TABLE_NAME + " LEFT JOIN " + BookmarkTagTable.TABLE_NAME +
          " ON " + BookmarksTable.TABLE_NAME + "." + BookmarksTable.ID + " = " +
          BookmarkTagTable.TABLE_NAME + "." + BookmarkTagTable.BOOKMARK_ID

    private const val CREATE_BOOKMARKS_TABLE =
      " create table if not exists " + BookmarksTable.TABLE_NAME + " (" +
          BookmarksTable.ID + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
          BookmarksTable.SURA + " INTEGER, " +
          BookmarksTable.AYAH + " INTEGER, " +
          BookmarksTable.PAGE + " INTEGER NOT NULL, " +
          BookmarksTable.ADDED_DATE +
          " TIMESTAMP DEFAULT CURRENT_TIMESTAMP);"

    private const val CREATE_TAGS_TABLE =
      " create table if not exists " + TagsTable.TABLE_NAME + " (" +
          TagsTable.ID + " INTEGER PRIMARY KEY, " +
          TagsTable.NAME + " TEXT NOT NULL, " +
          TagsTable.ADDED_DATE + " TIMESTAMP DEFAULT CURRENT_TIMESTAMP);"

    private const val CREATE_BOOKMARK_TAG_TABLE =
      " create table if not exists " + BookmarkTagTable.TABLE_NAME + " (" +
          BookmarkTagTable.ID + " INTEGER PRIMARY KEY, " +
          BookmarkTagTable.BOOKMARK_ID + " INTEGER NOT NULL, " +
          BookmarkTagTable.TAG_ID + " INTEGER NOT NULL, " +
          BookmarkTagTable.ADDED_DATE +
          " TIMESTAMP DEFAULT CURRENT_TIMESTAMP);"

    private const val BOOKMARK_TAGS_INDEX =
      "create unique index if not exists " +
          BookmarkTagTable.TABLE_NAME + "_index on " +
          BookmarkTagTable.TABLE_NAME + "(" +
          BookmarkTagTable.BOOKMARK_ID + "," +
          BookmarkTagTable.TAG_ID + ");"

    private const val LAST_PAGES_TABLE =
      "create table if not exists " + LastPagesTable.TABLE_NAME + " (" +
          LastPagesTable.ID + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
          LastPagesTable.PAGE + " INTEGER NOT NULL UNIQUE, " +
          LastPagesTable.ADDED_DATE + " TIMESTAMP DEFAULT CURRENT_TIMESTAMP);"

    private var sInstance: BookmarksDBHelper? = null

    fun getInstance(context: Context): BookmarksDBHelper {
      if (sInstance == null) {
        sInstance = BookmarksDBHelper(context.applicationContext)
      }
      return sInstance as BookmarksDBHelper
    }
  }

  private val lastPage: Int

  init {
    val quranSettings = QuranSettings.getInstance(context)
    lastPage = quranSettings.lastPage
  }

  override fun onCreate(db: SQLiteDatabase) {
    db.execSQL(CREATE_BOOKMARKS_TABLE)
    db.execSQL(CREATE_TAGS_TABLE)
    db.execSQL(CREATE_BOOKMARK_TAG_TABLE)
    db.execSQL(BOOKMARK_TAGS_INDEX)
    db.execSQL(LAST_PAGES_TABLE)
  }

  override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
    if (newVersion <= oldVersion) {
      Timber.w("Can't downgrade from version %d to version %d", oldVersion, newVersion)
      return
    }
    Timber.i("Upgrading database from version %d to version %d", oldVersion, newVersion)
    if (oldVersion < 2) {
      upgradeToVer2(db)
    }

    if (oldVersion < 3) {
      upgradeToVer3(db)
    }
  }

  private fun upgradeToVer3(db: SQLiteDatabase) {
    db.execSQL(LAST_PAGES_TABLE)

    // the default lastPage is -1
    if (lastPage >= Constants.PAGES_FIRST) {
      db.execSQL("INSERT INTO last_pages(page) values(?)", arrayOf<Any>(lastPage))
    }
  }

  private fun upgradeToVer2(db: SQLiteDatabase) {
    db.execSQL("DROP TABLE IF EXISTS tags")
    db.execSQL("DROP TABLE IF EXISTS ayah_tag_map")
    db.execSQL(CREATE_BOOKMARKS_TABLE)
    db.execSQL(CREATE_TAGS_TABLE)
    db.execSQL(CREATE_BOOKMARK_TAG_TABLE)
    db.execSQL(BOOKMARK_TAGS_INDEX)
    copyOldBookmarks(db)
  }

  private fun copyOldBookmarks(db: SQLiteDatabase) {
    try {
      // Copy over ayah bookmarks
      db.execSQL("INSERT INTO bookmarks(_id, sura, ayah, page) " +
          "SELECT _id, sura, ayah, page FROM ayah_bookmarks WHERE " +
          "bookmarked = 1")

      // Copy over page bookmarks
      db.execSQL("INSERT INTO bookmarks(page) " +
          "SELECT _id from page_bookmarks where bookmarked = 1")
    } catch (e: Exception) {
      Timber.e(e, "Failed to copy old bookmarks")
    }
  }
}
